<a href="https://colab.research.google.com/github/Likhitha33/DATA-ANALYSIS/blob/main/CodeforcesData.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import numpy as np
import pandas as pd
import requests
import json
from bs4 import BeautifulSoup 

**GETTING DATA FROM CODEFORCES API**

Data can be extracted in three forms.
1.   Using API key 
2.   Using requests library
3.   Using webscraping







First Using API key. Get the Api key from the website and store it in variable.

In [2]:
API_KEY='638414bd8870635963f080154602d9744894286b'

Using requests library. Get the Http request from the required website.We get the data in json format.We need to convert this data into pandas dataframe for further steps.

**CONTEST RATING DATA**

In [3]:
#Created Pandas Dataframe to store the contest rating data.
contest_rating_df=pd.DataFrame(columns=['contestId','handle_name','old_rating','new_rating','rank'])
contest_rating_df

Empty DataFrame
Columns: [contestId, handle_name, old_rating, new_rating, rank]
Index: []

Getting the codeforces rating data from the codeforces website.

In [ ]:
#Data of codeforces contestId 1
url_contest_rating='https://codeforces.com/api/contest.ratingChanges?contestId=1'
response_rating=requests.get(url_contest_rating)
response_rating.json()

Collected handle_name,old_rating,newrating,Rank from contestId 1 to 10  from contest rating.

In [5]:
#Getting the data of contestId from 1 to 10
for i in range(1,10):
  url_contest_rating1='https://codeforces.com/api/contest.ratingChanges?contestId='+str(i)
  response_cr=requests.get(url_contest_rating1)
  data=response_cr.json()
  #converting json data to pandas dataframe
  #Here we are only taking  attributes required for analysis
  for contest in  data['result']:
    contest_id=contest['contestId']
    handle_name=contest['handle']
    old_rating=contest['oldRating']
    new_rating=contest['newRating']
    rank=contest['rank']
    
    contest_rating_df=contest_rating_df.append({'contestId':contest_id,'handle_name':handle_name,'old_rating':old_rating,
                                              'new_rating':new_rating,'rank':rank},ignore_index=True)

In [6]:
contest_rating_df

contestId handle_name old_rating new_rating rank
0            1   vepifanov       1500       1600    1
1            1      Orfest       1500       1597    2
2            1        NALP       1500       1593    3
3            1      forest       1500       1590    4
4            1         rem       1500       1580    5
...        ...         ...        ...        ...  ...
2392         9        TSAM       1500       1386  251
2393         9     nep1965       1500       1386  251
2394         9  humaneitor       1500       1386  251
2395         9      Hernan       1500       1386  251
2396         9    Alisher2       1500       1386  251

[2397 rows x 5 columns]

The contest_rating_df has 2397 rows and 5 columns.

Now that we have the data in required format (pandas dataframe). We have to check whether the data is clean or not.

**DATA CLEANING**

Data cleaning is a part of the process on a data science project.
Data cleaning is the process of fixing or removing incorrect, corrupted, incorrectly formatted, duplicate, or incomplete
data within a dataset.
When you clean your data, all outdated or incorrect information is gone – leaving you with the highest quality information for your analysis and
model building.

In [7]:
contest_rating_df.dtypes

contestId      object
handle_name    object
old_rating     object
new_rating     object
rank           object
dtype: object

In [8]:
contest_rating_df.isna().sum()

contestId      0
handle_name    0
old_rating     0
new_rating     0
rank           0
dtype: int64

Thankfully data has no null values. If there are null values count is displayed and we have to make sure that data has no null values.

In [9]:
contest_rating_df=contest_rating_df.append({'contestId':np.nan,'handle_name':np.nan,'old_rating':np.nan,
                                              'new_rating':np.nan,'rank':np.nan},ignore_index=True)
contest_rating_df=contest_rating_df.append({'contestId':'2378','handle_name':'someHanldename','old_rating':np.nan,
                                              'new_rating':np.nan,'rank':np.nan},ignore_index=True)


In [10]:
contest_rating_df.tail(3)

contestId     handle_name old_rating new_rating rank
2396         9        Alisher2       1500       1386  251
2397       NaN             NaN        NaN        NaN  NaN
2398      2378  someHanldename        NaN        NaN  NaN

we can see that nan values are added to contest_rating_df.

In [11]:
contest_rating_df.isna().sum()

contestId      1
handle_name    1
old_rating     2
new_rating     2
rank           2
dtype: int64

In [12]:
contest_rating_df.dropna(how='any',inplace=True)

In [13]:
contest_rating_df.isna().sum()

contestId      0
handle_name    0
old_rating     0
new_rating     0
rank           0
dtype: int64

we can see that we have removed the null values.

In [14]:
contest_rating_df.duplicated

<bound method DataFrame.duplicated of      contestId handle_name old_rating new_rating rank
0            1   vepifanov       1500       1600    1
1            1      Orfest       1500       1597    2
2            1        NALP       1500       1593    3
3            1      forest       1500       1590    4
4            1         rem       1500       1580    5
...        ...         ...        ...        ...  ...
2392         9        TSAM       1500       1386  251
2393         9     nep1965       1500       1386  251
2394         9  humaneitor       1500       1386  251
2395         9      Hernan       1500       1386  251
2396         9    Alisher2       1500       1386  251

[2397 rows x 5 columns]>

In [15]:
contest_rating_df.drop_duplicates()

contestId handle_name old_rating new_rating rank
0            1   vepifanov       1500       1600    1
1            1      Orfest       1500       1597    2
2            1        NALP       1500       1593    3
3            1      forest       1500       1590    4
4            1         rem       1500       1580    5
...        ...         ...        ...        ...  ...
2392         9        TSAM       1500       1386  251
2393         9     nep1965       1500       1386  251
2394         9  humaneitor       1500       1386  251
2395         9      Hernan       1500       1386  251
2396         9    Alisher2       1500       1386  251

[2397 rows x 5 columns]

Stroing the pandas dataframe to excel.

In [16]:
contest_rating_df.to_excel("contest_rating.xlsx")

**CONTEST STATUS DATA**

In [ ]:
URL = 'https://codeforces.com/api/contest.status?contestId=1&from=1&count=10'
page = requests.get(URL)
soup = BeautifulSoup(page.content)
soup

In [19]:
type(soup)

bs4.BeautifulSoup

In [20]:
contest_status_df1=pd.DataFrame(columns=['contestId','handle_name','ProgrammingLanguage','Problem_Index','Problem_Name','Problem_tags'])
contest_status_df1

Empty DataFrame
Columns: [contestId, handle_name, ProgrammingLanguage, Problem_Index, Problem_Name, Problem_tags]
Index: []

In [21]:
contest_data=[]
OK = 200
for i in range(1,10):
  url_contest_status='https://codeforces.com/api/contest.status?contestId='+str(i)+'&from=1&count=10'
  response_status=requests.get(url_contest_status)
  if response_status.status_code == OK:
    conteststatus=response_status.json()
    contest_data+=conteststatus['result']
  else:
    print(i,"No Data",response_status.status_code)
    
else:
  print(contest_data)
  print("WORKED")

[{'id': 181885405, 'contestId': 1, 'creationTimeSeconds': 1669036872, 'relativeTimeSeconds': 2147483647, 'problem': {'contestId': 1, 'index': 'A', 'name': 'Theatre Square', 'type': 'PROGRAMMING', 'rating': 1000, 'tags': ['math']}, 'author': {'contestId': 1, 'members': [{'handle': 'BMS250'}], 'participantType': 'PRACTICE', 'ghost': False, 'startTimeSeconds': 1266580800}, 'programmingLanguage': 'Python 3', 'verdict': 'WRONG_ANSWER', 'testset': 'TESTS', 'passedTestCount': 3, 'timeConsumedMillis': 46, 'memoryConsumedBytes': 0}, {'id': 181885396, 'contestId': 1, 'creationTimeSeconds': 1669036866, 'relativeTimeSeconds': 2147483647, 'problem': {'contestId': 1, 'index': 'A', 'name': 'Theatre Square', 'type': 'PROGRAMMING', 'rating': 1000, 'tags': ['math']}, 'author': {'contestId': 1, 'members': [{'handle': 'technickhil'}], 'participantType': 'PRACTICE', 'ghost': False, 'startTimeSeconds': 1266580800}, 'programmingLanguage': 'GNU C++17', 'verdict': 'OK', 'testset': 'TESTS', 'passedTestCount': 2

In [22]:
for contest_status in contest_data:
  contest_id=contest_status['contestId']
  problem_index = contest_status['problem']['index']
  problem_name = contest_status['problem']['name']
  problem_tags = contest_status['problem']['tags']
  handle_name = contest_status['author']['members'][0]['handle']
  handle_programmingLanguage = contest_status['programmingLanguage']
  contest_status_df1=contest_status_df1.append({'contestId':contest_id,'handle_name':handle_name,'ProgrammingLanguage':handle_programmingLanguage,
                                              'Problem_Index':problem_index,'Problem_Name':problem_name,'Problem_tags':problem_tags},ignore_index=True)

In [ ]:
contest_status_df1

In [24]:
contest_status_df1.dtypes

contestId              object
handle_name            object
ProgrammingLanguage    object
Problem_Index          object
Problem_Name           object
Problem_tags           object
dtype: object

In [25]:
contest_status_df1.isna().sum()

contestId              0
handle_name            0
ProgrammingLanguage    0
Problem_Index          0
Problem_Name           0
Problem_tags           0
dtype: int64

In [26]:
contest_status_df1.isnull().sum()

contestId              0
handle_name            0
ProgrammingLanguage    0
Problem_Index          0
Problem_Name           0
Problem_tags           0
dtype: int64

In [ ]:
contest_status_df1.duplicated

In [ ]:
contest_status_df1.head(5)

In [28]:
contest_status_df1.drop_duplicates(subset=['contestId', 'handle_name','ProgrammingLanguage','Problem_Index','Problem_Name'], keep='last',inplace=True)

In [ ]:
contest_status_df1

In [32]:
contest_status_df1.to_excel('contest Status.xlsx')

Both contest rating and contest status cleaned data is stored in excel sheets, Futher these are used in analysis.